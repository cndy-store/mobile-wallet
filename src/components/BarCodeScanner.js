import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Button, StatusBar, StyleSheet, View, Text } from 'react-native';
import Dimensions from 'Dimensions';
import Camera from 'react-native-camera';
import CameraNotAuthorized from '../components/CameraNotAuthorized';
import CameraAuthorizationPending from '../components/CameraAuthorizationPending';

const { width, height } = Dimensions.get('window');

export class BarCodeScanner extends Component {
  constructor(props) {
    super(props);

    this.state = {
      error: null
    };

    this.errorTimeout = null;
    this.handleCancel = this.handleCancel.bind(this);
    this.handleBarCodeRead = this.handleBarCodeRead.bind(this);
    // necessary to debounce the QR code callback from camera
    this.foundCode = false;
  }

  componentWillUnmount() {
    this.clearTimeouts();
  }

  clearErrors() {
    this.setState({ error: null });
  }

  clearTimeouts() {
    if (this.errorTimeout) clearTimeout(this.errorTimeout);
  }

  handleCancel() {
    this.props.onCancel();
  }

  handleBarCodeRead(e) {
    if (this.foundCode) return;

    const decoded = this.props.decoder(e.data);

    if (!decoded) {
      this.setState({ error: 'Not a valid QR code' });
      this.errorTimeout = setTimeout(() => this.clearErrors(), 1000);
    } else {
      this.foundCode = true;
      this.clearErrors();
      this.props.onCodeScan(decoded);
    }
  }

  renderCamera() {
    return (
      <Camera
        style={styles.camera}
        type={Camera.constants.Type.back}
        barCodeTypes={[Camera.constants.BarCodeType.qr]}
        onBarCodeRead={this.handleBarCodeRead}
        permissionDialogTitle="Allow Camera Access"
        permissionDialogMessage="We need it to scan QR codes"
        notAuthorizedView={<CameraNotAuthorized />}
        pendingAuthorizationView={<CameraAuthorizationPending />}
      />
    );
  }

  renderError() {
    if (!this.state.error) return null;

    return (
      <View style={[styles.overlay, styles.error]}>
        <Text>{this.state.error}</Text>
      </View>
    );
  }

  render() {
    return (
      <View style={styles.container}>
        <StatusBar animated hidden />
        {this.renderCamera()}
        <View style={[styles.overlay, styles.topOverlay]}>
          <Text>Scan a QR Code</Text>
        </View>
        <View style={[styles.overlay, styles.bottomOverlay]}>
          <View>
            <Button title={'Abort'} onPress={this.handleCancel} />
          </View>
        </View>
        {this.renderError()}
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    alignContent: 'center',
    height
  },
  camera: {
    flex: 1,
    justifyContent: 'flex-end',
    alignItems: 'center',
    width
  },
  overlay: {
    flex: 1,
    position: 'absolute',
    backgroundColor: 'rgba(255,255,255, 0.5)',
    padding: 16,
    right: 0,
    left: 0,
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center'
  },
  error: {
    left: width / 4,
    right: width / 4,
    width: width / 2,
    height: width / 2,
    borderRadius: 2,
    borderWidth: 1,
    borderColor: '#ffffff',
    backgroundColor: 'rgba(255,0,0,0.4)'
  },
  topOverlay: {
    top: 0
  },
  bottomOverlay: {
    bottom: 0
  }
});

BarCodeScanner.propTypes = {
  onCancel: PropTypes.func.isRequired,
  onCodeScan: PropTypes.func.isRequired,
  decoder: PropTypes.func.isRequired
};

export default BarCodeScanner;
