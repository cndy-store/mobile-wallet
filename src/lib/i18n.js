import NativeLocale from 'react-native-locale';

const localeSettings = NativeLocale.constants();

export { localeSettings };
