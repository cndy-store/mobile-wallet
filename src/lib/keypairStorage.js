import * as Keychain from 'react-native-keychain';

const getItem = async _key => {
  const credentials = await Keychain.getGenericPassword();
  return Promise.resolve(credentials.password);
};

const setItem = (key, value) => {
  return Keychain.setGenericPassword(key, value);
};

const removeItem = _key => {
  return Keychain.resetGenericPassword();
};

export { getItem, setItem, removeItem };
