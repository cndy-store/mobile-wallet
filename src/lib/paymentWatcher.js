let timeout = null;
let pollingInterval = 3000;

const start = (dispatcher, loadFn, publicKey) => {
  if (timeout) {
    return;
  }

  timeout = window.setTimeout(() => {
    dispatcher(loadFn(publicKey))
      .then(() => {
        timeout = null;
        start(dispatcher, loadFn, publicKey);
      })
      .catch(error => {
        console.error(error);
      });
  }, pollingInterval);
};

const stop = () => {
  if (timeout) window.clearTimeout(timeout);
  timeout = null;
};

export { start, stop };
