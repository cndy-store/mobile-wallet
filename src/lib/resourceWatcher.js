let timeouts = {};
let pollingInterval = 3000;

const start = (name, dispatcher, loadFn, publicKey) => {
  if (timeouts[name]) {
    return;
  }

  timeouts[name] = window.setTimeout(() => {
    dispatcher(loadFn(publicKey))
      .then(() => {
        timeouts[name] = null;
        start(name, dispatcher, loadFn, publicKey);
      })
      .catch(error => {
        console.error(error);
      });
  }, pollingInterval);
};

const stop = name => {
  if (timeouts[name]) window.clearTimeout(timeouts[name]);
  timeouts[name] = null;
};

const startPaymentsWatcher = start.bind(null, 'payments');
const stopPaymentsWatcher = stop.bind(null, 'payments');

const startAcccountWatcher = start.bind(null, 'account');
const stopAccountWatcher = stop.bind(null, 'account');

export {
  startPaymentsWatcher,
  stopPaymentsWatcher,
  startAcccountWatcher,
  stopAccountWatcher
};
