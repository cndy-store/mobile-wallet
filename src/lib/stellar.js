import axios from 'axios';
import StellarSdk from '../vendor/stellar-base-0.7.7';
StellarSdk.Network.useTestNetwork();

const assetCode = 'CNDY';
const issuerPublicKey =
  'GCJKC2MI63KSQ6MLE6GBSXPDKTDAK43WR522ZYR3F34NPM7Z5UEPIZNX';

const asset = new StellarSdk.Asset(assetCode, issuerPublicKey);

const axiosInstance = axios.create({
  baseURL: 'https://horizon-testnet.stellar.org',
  timeout: 10 * 1000
});

module.exports = {
  axios: axiosInstance,
  asset,
  StellarSdk
};
