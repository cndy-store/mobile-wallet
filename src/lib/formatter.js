import { padEnd, reduceRight } from 'lodash';
import { localeSettings } from './i18n';

const delimiter = localeSettings.decimalSeparator;
const finalDelimiter = '.';
const maxPrecisionLength = 7;

const allowedChars = new RegExp(`^[0-9${delimiter}${finalDelimiter}]+`);
const delimiterSplit = new RegExp(`[${delimiter}${finalDelimiter}]+`);

const sanitizeOngoingAmountInput = input => {
  const onlyAllowedChars = input.match(allowedChars);
  if (!onlyAllowedChars) return '';
  if (onlyAllowedChars[0] !== input) return onlyAllowedChars[0];

  const [precision, ...rest] = input.split(delimiterSplit);
  if (!precision.length) return '';
  if (!rest.length) return precision;

  let scale = rest.join('');
  if (scale.length > maxPrecisionLength) {
    scale = scale.substr(0, maxPrecisionLength);
  }

  return [precision, scale].join(delimiter);
};

const ensureScale = input => {
  const [precision, scale] = input.split(delimiter);
  const paddedScale = padEnd(scale || '', 2, '0');

  return `${precision}${delimiter}${paddedScale}`;
};

const parseTransactionAmount = input => {
  const sanitized = sanitizeOngoingAmountInput(input);
  if (sanitized === '') return null;
  const withScale = ensureScale(sanitized);

  const withCorrectDelimiter = withScale.replace(delimiter, finalDelimiter);
  return withCorrectDelimiter;
};

const optimizeScale = scale => {
  const scaleNumbers = scale.split('');
  const neededNumbers = reduceRight(
    scaleNumbers,
    (acc, number) => {
      if (acc.length || number !== '0') {
        acc.push(number);
      }
      return acc;
    },
    []
  );

  return neededNumbers.reverse().join('');
};

// this method is to be used with normalized amount strings
const shortFormat = input => {
  const [precision, scale] = input.split(finalDelimiter);
  const optimizedScale = optimizeScale(scale);
  const paddedScale = padEnd(optimizedScale, 2, '0');

  return `${precision}${delimiter}${paddedScale}`;
};

export { sanitizeOngoingAmountInput, parseTransactionAmount, shortFormat };
