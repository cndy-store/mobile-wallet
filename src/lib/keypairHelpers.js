import { asset, StellarSdk } from '../lib/stellar';
import bip21 from 'bip21';

const { StrKey } = StellarSdk;

const scheme = asset.getCode().toLowerCase();
const secretScheme = `${scheme}secret`;

const isValidPublicKey = publicKey => StrKey.isValidEd25519PublicKey(publicKey);
const isValidSecret = secret => StrKey.isValidEd25519SecretSeed(secret);

const encodePublicKey = (publicKey, options = {}) => {
  if (!isValidPublicKey(publicKey)) return null;
  options.scheme = scheme;
  return bip21.encode(publicKey, options);
};

const decodePublicKey = url => {
  let decoded;
  try {
    decoded = bip21.decode(url, scheme);
  } catch (e) {
    return null;
  }

  const publicKey = decoded.address;
  if (!isValidPublicKey(publicKey)) return null;

  return {
    publicKey,
    options: decoded.options
  };
};

const encodeSecret = (secret, options = {}) => {
  if (!isValidSecret(secret)) return null;
  options.scheme = secretScheme;
  return bip21.encode(secret, options);
};

const decodeSecret = url => {
  let decoded;
  try {
    decoded = bip21.decode(url, secretScheme);
  } catch (e) {
    return null;
  }

  const secret = decoded.address;
  if (!isValidSecret(secret)) return null;

  return {
    secret,
    options: decoded.options
  };
};

export {
  isValidPublicKey,
  isValidSecret,
  encodePublicKey,
  decodePublicKey,
  encodeSecret,
  decodeSecret
};
