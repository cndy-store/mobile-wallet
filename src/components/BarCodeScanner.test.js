import React from 'react';
import { asset } from '../lib/stellar';
import { BarCodeScanner } from './BarCodeScanner.js';

import renderer from 'react-test-renderer';

it('renders correctly', () => {
  const tree = renderer.create(<BarCodeScanner />).toJSON();
  expect(tree).toMatchSnapshot();
});
