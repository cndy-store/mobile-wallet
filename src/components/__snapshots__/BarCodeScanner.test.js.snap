// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "alignContent": "center",
      "alignItems": "center",
      "flex": 1,
      "height": 1334,
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "left": 0,
          "paddingLeft": 10,
          "paddingRight": 10,
          "paddingTop": 15,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
      ]
    }
  >
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "flex-start",
          "backgroundColor": "transparent",
          "borderBottomWidth": null,
          "borderColor": "#007aff",
          "borderLeftWidth": null,
          "borderRadius": 5,
          "borderRightWidth": null,
          "borderTopWidth": null,
          "elevation": 0,
          "flexDirection": "row",
          "height": 45,
          "justifyContent": "space-between",
          "opacity": 1,
          "paddingBottom": 6,
          "paddingTop": 6,
          "shadowColor": null,
          "shadowOffset": null,
          "shadowOpacity": null,
          "shadowRadius": null,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "backgroundColor": "transparent",
            "color": "#007aff",
            "fontFamily": "System",
            "fontSize": 16.5,
            "marginLeft": 0,
            "marginRight": 0,
            "paddingLeft": 16,
            "paddingRight": 16,
          }
        }
        uppercase={false}
        virtual={undefined}
      >
        Cancel
      </Text>
    </View>
  </View>
</View>
`;
